Revised detection of arithmetic instructions with immediate operands

xavierleroy opened the pull request:
Many processor instructions support "immediate" (i.e. constant) arguments, encoded in the instruction itself.  Which instructions have immediate forms and which range of immediate integers are supported depend a lot on the target processor.

Curently, the generic instruction selection pass Selectgen assumes a simplified model where all integer arithmetic and logical operations support immediate arguments in the same range.  The range is given by the processor-specific `is_immediate` method.

This may have been adequate for the Alpha back in the 1990s, but today none of the OCaml target platforms fit this model.  Exceptions must be introduced as special cases in the processor-specific `select_operation` method.

This PR replaces the single `is_immediate n` method that is supposed to apply to all arithmetic instructions by two methods:
    
 -  `is_immediate op n` : tests whether `n` is in the range of supported immediate arguments for integer operation `op`
 -  `is_immediate_test cmp n` : tests whether `n` is in the range of supported immediate arguments for integer comparison `cmp`
 
This makes it possible to handle operations without immediate operands (e.g. multiply or multiply-high on many platforms) and operations with specific ranges of immediate operands (e.g. N-bit unsigned versus N-bit signed) without adding special cases in the `select_operation` method.  Processor-specific code is therefore significantly reduced (look at RISC-V for an example). 

There are two additional commits:
- Commit 979e9391b revises the handling of ADD/SUB/CMP immediate for ARM 64 bits.  In particular, a wider range of immediates is supported for ADD and SUB using two-instruction code sequences.
- Commit 2a6c09a8a simplifies a few integer constants that were obfuscated so as to pass compilation on a 32-bit host, as "make check_all_arches" would do if ran on a 32-bit host.  However, "make check_all_arches" does not run on 32-bit hosts, unlike what is claimed in comments.  More generally, 32-bit hosts are no longer used for developing OCaml and will not be used for cross-compilation.  So, let's not complicate the back-ends unnecessarily.

xavierleroy committed:
Revised detection of arithmetic instructions with immediate operands

Replace the a single `is_immediate n` method that is supposed to apply
to all arithmetic instructions by two methods:

`is_immediate op n` : tests whether `n` is in the range of supported
   immediate arguments for integer operation `op`
`is_immediate_test cmp n` : tests whether `n` is in the range of supported
   immediate arguments for integer comparison `cmp`

This makes it easier to handle operations without immediate operands
(e.g. multiply or multiply-high on many platforms) and operations with
specific ranges of immediate operands (e.g. N-bit unsigned versus
N-bit signed).  Before, these operations had to be treated as special
cases in the platform-specific `select_operation` method.

xavierleroy committed:
ARM64: revised generation of ADD/SUB/CMP immediate

- Support a wider range of immediate values for ADD and SUB,
  using two instructions when needed (add/sub middle 12 bits then low 12 bits).

- Do not rely on the assembler to convert CMP immediate negative to
  CMN immediate.

xavierleroy committed:
Back-ends for 64-bit platforms do not need to be compilable on a 32-bit host

This commit simplifies a few integer constants that were obfuscated so
as to pass compilation on a 32-bit host, as "make check_all_arches"
would do if ran on a 32-bit host.  However, "make check_all_arches"
does not run on 32-bit hosts, unlike what is claimed in comments.

More generally, 32-bit hosts are no longer used for developing OCaml and
will not be used for cross-compilation.  So, let's not complicate the
back-ends unnecessarily.

shindere commented:
Sorry I am not expert enough to review this very interesting PR
properly.

Perhaps @mshinwell could have a look ?

stedolan commented on asmcomp/amd64/selection.ml:
```diff
@@ -121,17 +121,23 @@ let inline_ops =
+let is_immediate n = n <= 0x7FFF_FFFF && n >= -0x8000_0000
+
+let is_immediate_natint n = n <= 0x7FFFFFFFn && n >= -0x80000000n
+
```
Purely stylistic suggestion to take or leave: I find this easier to read as:
```ocaml
let is_immediate n =
  Int32.to_int (Int32.of_int n) = n
```

stedolan commented on asmcomp/power/selection.ml:
```diff
@@ -43,13 +43,28 @@ let rec select_addr = function
+method is_immediate_test cmp n =
+  match cmp with
+  | Isigned _ -> is_immediate n
+  | Iunsigned _ -> is_immediate_logical n
```
I don't think the old code made this distinction between immediate ranges for signed and unsigned comparisons (or at least, I couldn't find it). Was the old code wrong?

xavierleroy replied:
Yes, it was wrong on the IBM platforms (Power and Z), which interpret immediate arguments differently whether the operation is signed or unsigned.  I think the code produced by Cmmgen never ran into the problematic case (using the immediate form of an unsigned operation when the argument is a small negative number).

stedolan commented on asmcomp/power/selection.ml:
```diff
@@ -43,13 +43,28 @@ let rec select_addr = function
+  | Iand | Ior | Ixor -> is_immediate_logical n
+  | Icomp c -> self#is_immediate_test c n
+  | Icheckbound _ -> is_immediate n
+  | _ -> assert false
```
I don't like the `_ -> assert false`, because it will remain exhaustive when a new instruction is added. Suggestion:

```ocaml
  | Ilsl | Ilsr | Iasr -> assert false (* Selectgen.select_shift *)
  | Idiv | Imod -> assert false        (* Selectgen: no immediates allowed *)
```

Alternatively, the `assert false` could be removed entirely by moving these two decisions into `is_immediate` rather than hardcoding them in Selectgen, now that `is_immediate` is able to make per-operation decisions.

xavierleroy replied:
None of the methods in $ARCH/selection.ml enumerates all cases explicitly: they all have a final catch-all case (which, generally, calls the method of the superclass).  That's in the spirit of this code: common behavior is implemented in selectgen.ml, and each $ARCH/selection.ml adds behavior only for the constructs that need special handling.

So, enumerating all cases here is not in the spirit of this code, I'm sorry.  But `_ -> false` could be a more appropriate default than `_ -> assert false`, just because it's always safe to not recognize an immediate argument.  Or `_ -> super#is_immediate c n` if we really want to be in the spirit.  This could also avoid special-casing shifts and divisions, as you remark.  Let me come back with an alternate proposal.

xavierleroy replied:
Commit f8eed137c reimplements the `is_immediate` method in the spirit I was trying to describe.  There's a fall-back case `_ -> super#is_immediate c n` that invokes a default `is_immediate` method that deals with shifts and returns `false` otherwise.  Each port lists the Mach operations for which it has immediate forms, and none other.  The default behavior is to return `false`, which may be suboptimal (a supported immediate form is not recognized) but safe.

stedolan replied:
This is nice, thanks!

stedolan commented on asmcomp/arm64/selection.ml:
```diff
@@ -96,10 +103,16 @@ class selector = object(self)
+  | Imul | Imulh -> false        (* no mul immediate instruction *)
+  | Iand | Ior | Ixor -> is_logical_immediate n
+  | Icomp _ | Icheckbound _ -> is_immediate n
+  | _ -> assert false
```
(same comment as power)

stedolan commented on asmcomp/arm64/emit.mlp:
```diff
@@ -318,6 +318,37 @@ let output_epilogue f =
+    `	sub	{emit_reg rd}, {emit_reg rs}, #{emit_int nh}\n`;
+    if nl <> 0 then
+      `	sub	{emit_reg rd}, {emit_reg rd}, #{emit_int nl}\n`
+  end
```
This code looks correct to me, but I'd feel better if there was a test for the new immediate codegen. (It wouldn't surprise me if there was nowhere in the testsuite that actually adds a constant above 4096 which is not divisible by 4096, in which case nothing in precheck would hit this path).

xavierleroy replied:
I'll think about a more general test for immediate operands, not restricted to add/sub for ARM64.

xavierleroy replied:
I've pushed a general test for immediate arguments.  Will check the "precheck" results tomorrow.

xavierleroy committed:
New test exercising immediate arguments to integer operations

stedolan commented on asmcomp/power/selection.ml:
```diff
@@ -43,13 +43,29 @@ let rec select_addr = function
+method is_immediate op n =
+  match op with
+  | Iadd | Imul -> is_immediate n
+  | Isub -> is_immediate (-n)  (* turned into add opposite *)
```
I think this may also need to be changed in the RISCV backend, which also does the `-n` trick.

(I think that it would be less error-prone to pick `Iadd` during instruction selection, rather than having the assembly emitter convert between add and sub)

xavierleroy replied:
> I think this may also need to be changed in the RISCV backend, which also does the -n trick.

Right.  I fixed that and tested on RISC-V.

> (I think that it would be less error-prone to pick Iadd during instruction selection, rather than having the assembly emitter convert between add and sub)

Maybe we've been through that already, but I like to leave open the possibility that a compilation pass will generate a Isub immediate after instruction selection.

stedolan replied:
Fair enough!

xavierleroy committed:
Changes for 9864

xavierleroy commented:
I'm still working on this one, but I pushed a cleaned-up version so that CI can run.

xavierleroy committed:
Test for immediates: also test "checkbound"

Plus: bow to check-typo.

xavierleroy committed:
Revised detection of arithmetic instructions with immediate operands, continued

- Rewrite the `is_immediate` methods in $ARCH/selection.ml in the style of
  other selection methods: operations that need platform-dependent handling
  are explicitly listed, all others fall through `super#is_immediate`.

- The `is_immediate` method from selectgen.ml knows how to handle shifts
  (and no other operation).  Remove the `select_shift_op` method,
  now unnecessary.

- ARM: remove special cases for multiply and multiply-high, no longer
  necessary.

- RISC-V: in emit.mlp, remove implementation of checkbound immediate,
  which is no longer generated.

xavierleroy commented:
After some refactoring and much testing, I think this PR is ready for the second round of reviewing.

stedolan commented on asmcomp/power/selection.ml:
```diff
@@ -43,13 +43,28 @@ let rec select_addr = function
+  | Isub -> is_immediate (-n)  (* turned into add opposite *)
+  | Iand | Ior | Ixor -> is_immediate_logical n
+  | Icomp c -> self#is_immediate_test c n
+  | Icheckbound _ -> is_immediate_logical n   (* unsigned comparison *)
```
I think is `is_immediate` was better than `is_immediate_logical`, although neither is quite right.

The `tdlle` / `twlle` instruction that implements checkbound takes a 16-bit immediate argument. The comparison is unsigned, but the immediate is sign-extended (!). So the representable immediates are (on 64-bit):
```
0 <= n <= 0x7FFF
0xFFFF_FFFF_FFFF_8000 <= n <= 0xFFFF_FFFF_FFFF_FFFF
```
Since there can't be arrays long enough to hit the second case, `is_immediate` probably does the right thing, but `0 <= n <= 0x7FFF` seems like the right test to put here.

xavierleroy replied:
Right, `is_immediate_logical` causes errors in testing, I thought I already reverted to `is_immediate`.  But your suggestion `0 <= n <= 0x7FFF` is better, I'll implement this ASAP.

stedolan reviewed (approved):
I have run out of nits to pick! (looks good to merge, assuming CI agrees)
